elixir/apps/domain/lib/domain/config/configuration/clients_upstream_dns.ex

```diff
@@ -0,0 +1,61 @@
+defmodule Domain.Config.Configuration.ClientsUpstreamDNS do
+  @moduledoc """
+  Embedded Schema for Clients Upstream DNS
+  """
+  use Domain, :schema
+  import Domain.Validator
+  import Ecto.Changeset
+  alias Domain.Types.IPPort
+
+  @primary_key false
+  embedded_schema do
+    field :protocol, Ecto.Enum, values: [:ip_port, :dns_over_tls, :dns_over_http]
+    field :address, :string
+  end
+
+  def changeset(dns_config \\ %__MODULE__{}, attrs) do
+    dns_config
+    |> cast(attrs, [:protocol, :address])
+    |> validate_required([:protocol, :address])
+    |> trim_change(:address)
+    |> validate_inclusion(:protocol, supported_protocols(),
+      message: "this type of DNS provider is not supported yet"
+    )
+    |> validate_address()
+  end
+
+  def supported_protocols do
+    ~w[ip_port]a
+  end
+
+  def validate_address(changeset) do
+    if has_errors?(changeset, :protocol) do
+      changeset
+    else
+      case fetch_field(changeset, :protocol) do
+        {_, :ip_port} -> validate_ip_port(changeset)
```

Comment:
Will just make it a bit more explicit
```suggestion
        {_changes_or_data, :ip_port} -> validate_ip_port(changeset)
```

Reply:
Sounds good 👍

Reply:
This has been updated.